airbyte-integrations/bases/base-normalization/normalization/transform_catalog/stream_processor.py

```diff
@@ -0,0 +1,527 @@
+        self.raw_schema = raw_schema
+        self.schema = schema
+        self.json_column_name = json_column_name
+        self.properties = properties
+        self.tables_registry = tables_registry
+
+        self.name_transformer = DestinationNameTransformer(integration_type)
+        self.json_path = [stream_name]
+        return self
+
+    def process(self, from_table: str) -> Dict:
+        """
+        @param from_table refers to the raw source table to use to extract data from
+        """
+        # Check properties
+        if not self.properties:
+            print(f"  Ignoring substream '{self.stream_name}' from {self.current_json_path()} because properties list is empty")
+            return
+
+        # Transformation Pipeline for this stream
+        from_table = self.write_model(self.generate_json_parsing_model(from_table), is_intermediate=True, suffix="ab1")
+        from_table = self.write_model(self.generate_column_typing_model(from_table), is_intermediate=True, suffix="ab2")
+        from_table = self.write_model(self.generate_id_hashing_model(from_table), is_intermediate=True, suffix="ab3")
+        from_table = self.write_model(self.generate_final_model(from_table), is_intermediate=False)
+        return {from_table: self.find_children_streams()}
+
+    def find_children_streams(self) -> List:
+        properties = self.properties
+        children: List[StreamProcessor] = []
+        for field in properties.keys():
+            children_properties = None
+            if is_airbyte_column(field):
+                pass
+            elif is_combining_node(properties[field]):
+                # TODO: merge properties of all combinations
+                pass
+            elif "type" not in properties[field] or is_object(properties[field]["type"]):
+                children_properties = find_properties_object([], field, properties[field])
+                is_nested_array = False
+                json_column_name = f"'{field}'"
+            elif is_array(properties[field]["type"]) and "items" in properties[field]:
+                quoted_field = self.name_transformer.normalize_column_name(field, in_jinja=True)
+                children_properties = find_properties_object([], field, properties[field]["items"])
+                is_nested_array = True
+                json_column_name = f"unnested_column_value({quoted_field})"
+            if children_properties:
+                for child_key in children_properties:
+                    stream_processor = StreamProcessor().initFromParent(
+                        parent=self,
+                        child_name=field,
+                        json_column_name=json_column_name,
+                        properties=children_properties[child_key],
+                        is_nested_array=is_nested_array,
+                    )
+                    children.append(stream_processor)
+        return children
+
+    def generate_json_parsing_model(self, from_table: str) -> str:
```

Comment:
this is cool. to make sure i understand. if i had the object: `{ name: "charles", preferences: { color: "blue", fruit: "apple" }}` this is essentially doing `SELECT name, preferences` at the top level and then one level down it would do `SELECT preferences.color, preferences.apple` using whatever json extraction syntax is appropriate?

Reply:
Yes, it creates two queries, since the `raw_source` table stores data in `_airbyte_data` column:
- top-level will do: `select _airbyte_data.name, _airbyte_data.preferences from raw_source`
- and a second based on the previous result: `select preferences.color, preferences.fruit from previous_table`

(just with the right json_extract function calls that makes this SQL a little more complicated to read)

Reply:
nice. i think it would be helpful to have this documented. i think this is just kinda complicated and the code will be easier to read if the reader knows what these things are trying to acheive.